src/codegen/codegen-match.lisp

```diff
@@ -0,0 +1,320 @@
+      ((not (and
+             (equalp (ast:node-type (ast:node-match-expr match)) tc:*boolean-type*)
+             (= 2 (length branches))))
+       (values nil nil nil))
+
+      ;; Case #2:
+      ;;
+      ;; Emit `cl:if' when first branch is `True' and second branch is `False'
+      ((and (equalp true-pattern (ast:match-branch-pattern (first branches)))
+            (equalp false-pattern (ast:match-branch-pattern (second branches))))
+       (values t
+               (ast:match-branch-body (first branches))
+               (ast:match-branch-body (second branches))))
+
+      ;; Case #3:
+      ;;
+      ;; Emit `cl:if' when second branch is `True' and first branch is `False'
+      ((and (equalp true-pattern (ast:match-branch-pattern (second branches)))
+            (equalp false-pattern (ast:match-branch-pattern (first branches))))
+       (values t
+               (ast:match-branch-body (second branches))
+               (ast:match-branch-body (first branches))))
+
+      ;; Case #4:
+      ;;
+      ;; Do not emit `cl:if' unless exhaustively matching boolean with
+      ;; constructor pattern branches.
+      (t
+       (values nil nil nil)))))
+
+;;;
+;;; Codegen Functions
+;;;
+
+(defun codegen-binding-types (bindings types)
+  "Generate type declarations for match bindings."
+  (declare (type list bindings)
+           (type t types)
+           (values t &optional))
+
+  `(declare
+    (ignorable ,@(mapcar #'car bindings))
+    ,@(if (not settings:*emit-type-annotations*)
+          nil
+          (loop :for binding :in bindings
+                :for var := (car binding)
+                :for type :in types
+                :collect `(type ,type ,var)))))
+
+(defun codegen-cond-match-branch (code pattern match-var match-expr-type env)
+  "Generate code for a `cl:cond' branch."
+  (declare (type t code)
+           (type pattern:pattern pattern)
+           (type symbol match-var)
+           (type tc:ty match-expr-type)
+           (type tc:environment env)
+           (values t &optional))
+
+  (multiple-value-bind (pred bindings types)
+      (codegen-pattern:codegen-pattern pattern match-var match-expr-type env)
+
+    `(,pred
+      ,(if (null bindings)
+           code
+           `(let ,bindings
+              ,(codegen-binding-types bindings types)
+              ,code)))))
+
+(defun codegen-case-match-branch (code pattern match-var match-expr-type env)
+  "Generate code for a `cl:case' branch."
+  (declare (type t code)
+           (type pattern:pattern pattern)
+           (type symbol match-var)
+           (type tc:ty match-expr-type)
+           (type tc:environment env)
+           (values t &optional))
+
+  (multiple-value-bind (pred bindings types)
+      (codegen-pattern:codegen-pattern pattern match-var match-expr-type env)
+    (declare (ignore pred))
+
+    (cond
+      ;; Case #1:
+      ;;
+      ;; Emit a list of a single literal pattern to match on.
+      ((pattern:pattern-literal-p pattern)
```

Comment:
I haven't looked so carefully. If we have a string literal, will this proc?

```
(case x
  (("string") ...))
```

won't work, because strings are not `eql`.

I think the answer is "no" because `case` should only work with enumeration types in this PR. But then I wouldn't expect a `pattern-literal` to exist. So maybe that should become an `assert`.